bookkeeper-server/src/main/java/org/apache/bookkeeper/client/RackawareEnsemblePlacementPolicyImpl.java

```diff
@@ -414,21 +414,33 @@ protected PlacementResult<List<BookieId>> newEnsembleInternal(
                 }
                 return PlacementResult.of(addrs, PlacementPolicyAdherence.FAIL);
             }
-
+            //Choose different rack nodes.
+            String curRack = null;
             for (int i = 0; i < ensembleSize; i++) {
-                String curRack;
                 if (null == prevNode) {
                     if ((null == localNode) || defaultRack.equals(localNode.getNetworkLocation())) {
                         curRack = NodeBase.ROOT;
                     } else {
                         curRack = localNode.getNetworkLocation();
                     }
                 } else {
-                    curRack = "~" + prevNode.getNetworkLocation();
+                    if (!curRack.startsWith("~")) {
+                        curRack = "~" + prevNode.getNetworkLocation();
+                    } else {
+                        curRack = curRack + NetworkTopologyImpl.NODE_SEPARATOR + prevNode.getNetworkLocation();
+                    }
+                }
+                try {
+                    prevNode = selectRandomFromRack(curRack, excludeNodes, ensemble, ensemble);
```

Comment:
Use `selectRandomFromRack` instead of `selectFromNetworkLocation`.
There are two reasons:
1. `selectFromNetworkLocation` params `fallbackToRandom` may step down to pick a node randomly.
2. If the user config `setNetworkTopologyStabilizePeriodSeconds`, RackawareEnsemblePlacementPolicy#selectFromNetworkLocation may step down to pick node from slave.

We should avoid it, just pick different node without steping down.